sensing/pointcloud_preprocessor/src/downsample_filter/voxel_grid_downsample_filter_nodelet.cpp

```diff
@@ -64,31 +64,144 @@ VoxelGridDownsampleFilterComponent::VoxelGridDownsampleFilterComponent(
+
+  // Set up the division multiplier
+  Eigen::Vector3i div_b_mul(1, div_b[0], div_b[0] * div_b[1]);
+
+  // Storage for mapping voxel coordinates to centroids
+  std::unordered_map<uint32_t, Centroid> voxel_centroid_map;
+  for (size_t global_offset = 0; global_offset + input->point_step <= input->data.size();
+       global_offset += input->point_step) {
+    Eigen::Vector3f point(
+      *reinterpret_cast<const float *>(&input->data[global_offset + x_offset]),
+      *reinterpret_cast<const float *>(&input->data[global_offset + y_offset]),
+      *reinterpret_cast<const float *>(&input->data[global_offset + z_offset]));
+    if (std::isfinite(point[0]) && std::isfinite(point[1]), std::isfinite(point[2])) {
+      int ijk0 = static_cast<int>(
+        std::floor(point[0] * inverse_voxel_size_[0]) - static_cast<float>(min_b[0]));
+      int ijk1 = static_cast<int>(
+        std::floor(point[1] * inverse_voxel_size_[1]) - static_cast<float>(min_b[1]));
+      int ijk2 = static_cast<int>(
+        std::floor(point[2] * inverse_voxel_size_[2]) - static_cast<float>(min_b[2]));
+      uint32_t voxel_id = ijk0 * div_b_mul[0] + ijk1 * div_b_mul[1] + ijk2 * div_b_mul[2];
+
+      if (voxel_centroid_map.find(voxel_id) == voxel_centroid_map.end()) {
+        voxel_centroid_map[voxel_id] = Centroid(point[0], point[1], point[2]);
+      } else {
+        voxel_centroid_map[voxel_id].add_point(point[0], point[1], point[2]);
+      }
+    }
+  }
+
+  // Copy the centroids to the output
+  output.row_step = voxel_centroid_map.size() * input->point_step;
+  output.data.resize(output.row_step);
+  size_t output_data_size = 0;
+  for (auto & pair : voxel_centroid_map) {
+    pair.second.calc_centroid();
+    if (transform_info.need_transform) {
+      Eigen::Array4f point = transform_info.eigen_transform *
+                             Eigen::Vector4f(pair.second.x, pair.second.y, pair.second.z, 1);
+      *reinterpret_cast<float *>(&output.data[output_data_size + x_offset]) = point[0];
+      *reinterpret_cast<float *>(&output.data[output_data_size + y_offset]) = point[1];
+      *reinterpret_cast<float *>(&output.data[output_data_size + z_offset]) = point[2];
+    } else {
+      *reinterpret_cast<float *>(&output.data[output_data_size + x_offset]) = pair.second.x;
+      *reinterpret_cast<float *>(&output.data[output_data_size + y_offset]) = pair.second.y;
+      *reinterpret_cast<float *>(&output.data[output_data_size + z_offset]) = pair.second.z;
+    }
+
+    *reinterpret_cast<float *>(&output.data[output_data_size + intensity_offset]) = 1.0f;
+    output_data_size += input->point_step;
+  }
+
+  // Post processing
+  pcl_conversions::fromPCL(xyz_fields_, output.fields);
+  output.width = voxel_centroid_map.size();
+  output.is_dense = true;  // we filter out invalid points
+  output.height = input->height;
+  output.is_bigendian = input->is_bigendian;
+  output.point_step = input->point_step;
```

Comment:
Can you separate a function for each processing unit?

Readability is improved when it is easy to see the following, and it is possible to replace pcl and this implementation.

```
  filter.setInputCloud(pointcloud_input);
  filter.setLeafSize(voxel_size_x_, voxel_size_y_, voxel_size_z_);
  filter.filter(pointcloud_output);
```